### PR TITLE
MULE-9392: Support configurable ObjectStore in every mule component

### PR DESCRIPTION
**Pull request by elrodro83**

### CODE REVIEW

File: core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java [hunk 1]
```diff
@@ -102,25 +103,29 @@ public void initialise() throws InitialisationException
...
     }
 
-    private ObjectStore<AtomicInteger> createStore() throws InitialisationException
+    private ObjectStore<AtomicInteger> acquireObjectStore()
     {
-        ObjectStoreManager objectStoreManager = (ObjectStoreManager) muleContext.getRegistry().get(
-                                MuleProperties.OBJECT_STORE_MANAGER);
-        return objectStoreManager.getObjectStore(flowConstruct.getName() + "." + getClass().getName(), false, -1,  60 * 5 * 1000, 6000 );
+        ObjectStore<AtomicInteger> os = this.store;
+        if (os == null)
```
Comment (pablokraan) [inline core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java]: Rename os variable

Comment (pablokraan) [inline core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java]: Be sure that this is not broken is object is initialized twice

File: core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java [hunk 2]
```diff
@@ -320,6 +325,21 @@ public void setIdExpression(String idExpression)
...
+    {
+        return store;
+    }
+
+    public void setObjectStore(ObjectStore<AtomicInteger> store)
+    {
+        this.store = store;
+    }
+
+    public boolean isInternalStore()
```
Comment (pablokraan) [inline core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java]: Why is this method needed?

File: core/src/main/java/org/mule/routing/AbstractAggregator.java [hunk 3]
```diff
@@ -77,6 +108,72 @@ public void initialise() throws InitialisationException
         eventCorrelator.setFailOnTimeout(isFailOnTimeout());
     }
 
+    protected ObjectStore<Long> acquireExpiredAndDispatchedGroupsStore(boolean persistentStores, String storePrefix)
+    {
+        ObjectStore<Long> os = expiredAndDispatchedGroupsObjectStore;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/AbstractAggregator.java]: Rename os variable

File: core/src/main/java/org/mule/routing/AbstractAggregator.java [hunk 4]
```diff
@@ -77,6 +108,72 @@ public void initialise() throws InitialisationException
         eventCorrelator.setFailOnTimeout(isFailOnTimeout());
     }
 
+    protected ObjectStore<Long> acquireExpiredAndDispatchedGroupsStore(boolean persistentStores, String storePrefix)
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/AbstractAggregator.java]: Why not returning a ListableObjectStore?

File: core/src/main/java/org/mule/routing/AbstractAggregator.java [hunk 5]
```diff
@@ -77,6 +108,72 @@ public void initialise() throws InitialisationException
...
+    {
+        ObjectStore<Long> os = expiredAndDispatchedGroupsObjectStore;
+        if (os == null)
+        {
+            internalExpiredAndDispatchedGroupsStore = true;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/AbstractAggregator.java]: Makes sense to have an object to contain each pair of flag/objectStore?

File: core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java [hunk 6]
```diff
@@ -320,6 +325,21 @@ public void setIdExpression(String idExpression)
         this.idExpression = idExpression;
     }
     
+    public ObjectStore<AtomicInteger> getObjectStore()
```
Comment (pablokraan) [inline core/src/main/java/org/mule/processor/IdempotentRedeliveryPolicy.java]: Maybe all these objectStore setters should say in the javadoc that the returned value can be different depending if the instance is initialized or not.

File: core/src/main/java/org/mule/routing/AbstractAggregator.java [hunk 7]
```diff
@@ -162,9 +306,34 @@ public void setStorePrefix(String storePrefix)
...
+        }
+        if (internalEventsObjectStore)
+        {
+            disposeIfDisposable(eventsObjectStore);
+        }
         eventCorrelator.dispose();
     }
 
+
+    private void disposeIfDisposable(Object o)
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/AbstractAggregator.java]: Why object?
Rename o parameter

File: core/src/main/java/org/mule/routing/AbstractCorrelationAggregator.java [hunk 8]
```diff
@@ -31,7 +31,7 @@ protected EventCorrelatorCallback getCorrelatorCallback(MuleContext muleContext)
     {
         public DelegateCorrelatorCallback(MuleContext muleContext)
         {
-            super(muleContext, persistentStores, storePrefix);
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/AbstractCorrelationAggregator.java]: Why is persistentStores attribute removed from the constructor?

File: core/src/main/java/org/mule/routing/EventGroup.java [hunk 9]
```diff
@@ -491,22 +496,19 @@ private void addAndOverrideSessionProperties(MuleSession session, MuleEvent even
...
         this.muleContext = context;
+    }
 
-        String storeKey = storePrefix + ".eventGroup." + groupId;
-        this.events = getObjectStoreManager().getObjectStore(storeKey, true);
+    private PartitionableObjectStore<MuleEvent> getEvents() throws ObjectStoreException
+    {
+        if (events == null)
+        {
+            events = muleContext.getRegistry().get(EVENTS_STORE_REGISTRY_KEY_PREFIX + storePrefix + ".eventGroup");
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/EventGroup.java]: I'm not sure that the objectStore will be always available on the registry. Can you check with LG?

File: core/src/main/java/org/mule/routing/correlation/EventCorrelator.java [hunk 10]
```diff
@@ -125,21 +120,13 @@ public EventCorrelator(EventCorrelatorCallback callback,
...
-                storePrefix + ".expiredAndDispatchedGroups", persistentStores);
-        processedGroups = (ListableObjectStore<Long>) objectStoreManager.getObjectStore(storePrefix
-                                                                                        + ".processedGroups",
-                                                                                        persistentStores, MAX_PROCESSED_GROUPS, -1, 1000);
-        eventGroups = (ListableObjectStore<EventGroup>) objectStoreManager.getObjectStore(storePrefix
-                                                                                          + ".eventGroups",
-                                                                                          persistentStores);
         this.flowConstruct = flowConstruct;
+
+        this.expiredAndDispatchedGroups = expiredAndDispatchedGroups;
```
Comment (pablokraan) [inline core/src/main/java/org/mule/routing/correlation/EventCorrelator.java]: In the previous version, EventCorrelator always created internal objectStores. In the new version, it always receives them injected. Why?

File: core/src/test/java/org/mule/routing/EventGroupTestCase.java [hunk 11]
```diff
@@ -6,28 +6,42 @@
...
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
-
 public class EventGroupTestCase extends AbstractMuleContextTestCase
 {
+    @Before
+    public void before() throws RegistrationException
+    {
+        ObjectStore<Serializable> os = muleContext.getRegistry().lookupObject(DefaultObjectStoreFactoryBean.class).createDefaultInMemoryObjectStore();
```
Comment (pablokraan) [inline core/src/test/java/org/mule/routing/EventGroupTestCase.java]: Rename os variable

File: modules/devkit-support/src/main/java/org/mule/security/oauth/RefreshTokenManager.java [hunk 12]
```diff
@@ -29,4 +31,15 @@
      */
     public void setMinRefreshIntervalInMillis(int minRefreshIntervalInMillis);
 
+    /**
+     * Sets the {@link ObjectStore} to use for the tokens refresh state.
+     * <p/>
+     * This can only be set during initialization. If this is set, the value given to
```
Comment (pablokraan) [inline modules/devkit-support/src/main/java/org/mule/security/oauth/RefreshTokenManager.java]: Before initialization

File: modules/spring-config/src/main/resources/META-INF/mule.xsd [hunk 13]
```diff
@@ -6118,17 +6125,47 @@
                         </xsd:documentation>
                     </xsd:annotation>
                 </xsd:attribute>
+                
+                <xsd:attribute name="expired-and-dispatched-groups-object-store-ref" type="xsd:string" use="optional">
```
Comment (pablokraan) [inline modules/spring-config/src/main/resources/META-INF/mule.xsd]: Not sure if thee is other way to do this, but having all these object stores in the XML config does not seem user friendly. Could we use a partitionable object store and create partitions for each different task?

File: tests/integration/src/test/java/org/mule/test/routing/CollectionAggregatorRouterCustomStoreTestCase.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return "collection-aggregator-router-custom-store.xml";
+    }
+
+    @Test
+    public void eventGroupWithCustomStore() throws Exception
+    {
+        MuleClient client = muleContext.getClient();
+        List<String> list = Arrays.asList("first", "second");
+        client.dispatch("vm://splitter", list, null);
```
Comment (pablokraan) [inline tests/integration/src/test/java/org/mule/test/routing/CollectionAggregatorRouterCustomStoreTestCase.java]: Don't use VM transport

File: core/src/main/java/org/mule/routing/AbstractCorrelationAggregator.java (hunk 8 above)
Comment (elrodro83) [inline core/src/main/java/org/mule/routing/AbstractCorrelationAggregator.java]: It was used to create the store for the groups, now it is looked up from the registry after being put there by the EventCorrelator constructor.

File: core/src/main/java/org/mule/routing/EventGroup.java (hunk 9 above)
Comment (elrodro83) [inline core/src/main/java/org/mule/routing/EventGroup.java]: Will check

**Pull request merged**